dialogue-core/src/main/java/com/palantir/dialogue/core/CautiousIncreaseAggressiveDecreaseConcurrencyLimiter.java

```diff
@@ -76,7 +82,62 @@
         return Optional.empty();
     }
 
-    final class Permit implements FutureCallback<Response> {
+    enum Behavior {
+        HOST_LEVEL() {
+            @Override
+            void onSuccess(Response result, PermitControl control) {
+                if (Responses.isInternalServerError(result) || Responses.isTooManyRequests(result)) {
+                    control.ignore();
+                } else if ((Responses.isQosStatus(result) && !Responses.isTooManyRequests(result))
+                        || Responses.isServerErrorRange(result)) {
+                    control.dropped();
+                } else {
+                    control.success();
+                }
+            }
+
+            @Override
+            void onFailure(Throwable throwable, PermitControl control) {
+                if (throwable instanceof IOException) {
+                    control.dropped();
+                } else {
+                    control.ignore();
+                }
+            }
+        },
+        ENDPOINT_LEVEL() {
+            @Override
+            void onSuccess(Response result, PermitControl control) {
+                if (Responses.isTooManyRequests(result) || Responses.isInternalServerError(result)) {
+                    control.dropped();
+                } else if (Responses.isServerErrorRange(result)) {
+                    control.ignore();
+                } else {
+                    control.success();
+                }
+            }
+
+            @Override
+            void onFailure(Throwable _throwable, PermitControl control) {
+                control.ignore();
+            }
+        };
+
+        abstract void onSuccess(Response result, PermitControl control);
+
+        abstract void onFailure(Throwable throwable, PermitControl control);
+    }
+
+    interface PermitControl {
+
+        void ignore();
+
+        void dropped();
+
+        void success();
+    }
```

Comment:
I know this terminology matches the netflix lib, but given how much indirection we already have, could we name them `unchangedLimit / decreaseLimit / increaseLimit` ?

Reply:
That sound reasonable, but I’m not sure I’d like to rename methods in this refactor as it’s already fairly large. This only extracts an interface from existing code.